honeybee_energy/programtype.py

```diff
@@ -63,5 +82,567 @@ def name(self):
+                                    t_people_mtx[1], timestep_resolution)
+        lighting = None
+        if len(lighting_mtx) != 0:
+            t_lighting_mtx = zip(*lighting_mtx)
+            lighting = Lighting.average('{}_Lighting'.format(name), t_lighting_mtx[0],
+                                        t_lighting_mtx[1], timestep_resolution)
+        electric_equipment = None
+        if len(e_equip_mtx) != 0:
+            t_e_equip_mtx = zip(*e_equip_mtx)
+            electric_equipment = ElectricEquipment.average(
+                '{}_Electric Equipment'.format(name), t_e_equip_mtx[0],
+                t_e_equip_mtx[1], timestep_resolution)
+        gas_equipment = None
+        if len(g_equip_mtx) != 0:
+            t_g_equip_mtx = zip(*g_equip_mtx)
+            gas_equipment = GasEquipment.average(
+                '{}_Gas Equipment'.format(name), t_g_equip_mtx[0],
+                t_g_equip_mtx[1], timestep_resolution)
+        infiltration = None
+        if len(inf_mtx) != 0:
+            t_inf_mtx = zip(*inf_mtx)
+            infiltration = Infiltration.average(
+                '{}_Infiltration'.format(name), t_inf_mtx[0],
+                t_inf_mtx[1], timestep_resolution)
+        ventilation = None
+        if len(vent_mtx) != 0:
+            t_vent_mtx = zip(*vent_mtx)
+            ventilation = Ventilation.average(
+                '{}_Ventilation'.format(name), t_vent_mtx[0],
+                t_vent_mtx[1], timestep_resolution)
+        setpoint = None
+        if len(setp_mtx) != 0:
+            t_setp_mtx = zip(*setp_mtx)
+            setpoint = Setpoint.average('{}_Setpoint'.format(name), t_setp_mtx[0],
+                                        t_setp_mtx[1], timestep_resolution)
+
+        # return the averaged object
+        return ProgramType(name, people, lighting, electric_equipment, gas_equipment,
+                           infiltration, ventilation, setpoint)
+
+    def duplicate(self):
+        """Get a copy of this object."""
+        return self.__copy__()
+
+    def lock(self):
+        """The lock() method to will also lock the loads."""
+        self._locked = True
+        if self.people is not None:
+            self.people._locked = True
+        if self.lighting is not None:
+            self.lighting._locked = True
+        if self.electric_equipment is not None:
+            self.electric_equipment._locked = True
+        if self.gas_equipment is not None:
+            self.gas_equipment._locked = True
+        if self.infiltration is not None:
+            self.infiltration._locked = True
+        if self.ventilation is not None:
+            self.ventilation._locked = True
+        if self.setpoint is not None:
+            self.setpoint._locked = True
+
+    def unlock(self):
+        """The unlock() method will also unlock the loads."""
+        self._locked = False
+        if self.people is not None:
+            self.people._locked = False
```

Comment:
calling the private methods outside the class itself is not good practice. I have seen better links but this can do (https://www.oreilly.com/ideas/how-to-make-mistakes-in-python). Search for `private`. This can result in bugs if at some point one changes `unlock` method and does more than just changing `_locked` to False.

Reply:
Fair point.  I'll go through and fix all of the places where I have custom `lock` and `unlock` methods that are set up like this and have them use the public method instead. For clarification, though, I would draw a distinction between "calling a private method" and "accessing a private attribute" since `_locked` is not a method.   I would say that you're right that you should never set a private variable outside of the module where it is introduced (so what I'm doing here is wrong). However, checking what a private attribute is (in some cases) very helpful and is often necessary when the hidden attribute can differ from the public one.

Reply:
> However, checking what a private attribute is (in some cases) very helpful and is often necessary when the hidden attribute can differ from the public one.

This is the original link that I was looking for. See their suggestion for resolving such cases: https://docs.quantifiedcode.com/python-anti-patterns/correctness/accessing_a_protected_member_from_outside_the_class.html

I agree that there might be very specific cases but generally speaking I would say access outside the _class_ is not a good practice.